test/xpu_api/utilities/optional/optional.object/optional.object.swap/swap_optional_object_swap.pass.cpp

```diff
@@ -0,0 +1,239 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include <oneapi/dpl/optional>
+#include <oneapi/dpl/type_traits>
+#include <oneapi/dpl/utility>
+
+#include "support/test_macros.h"
+#include "support/utils.h"
+
+
+using dpl::optional;
+
+class X
+{
+    int i_;
+
+  public:
+    X(int i) : i_(i) {}
+    X(X&& x) = default;
+    X&
+    operator=(X&&) = default;
+    ~X() {}
+
+    friend bool
+    operator==(const X& x, const X& y)
+    {
+        return x.i_ == y.i_;
+    }
+};
+
+class Y
+{
+    int i_;
+
+  public:
+    Y(int i) : i_(i) {}
+    Y(Y&&) = default;
```

Comment:
If we want to check that optional performs an ADL lookup for real object swapping, I guess we need to mark move constructor and move assignment operators as deleted to disable `std::swap` for two objects of `Y`.
Otherwise, nothing is tested by using `Y`.

Similar comment should be applied to `optional.specalg/swap_optional_specalg.pass.cpp`.

These tests also can be merged with each other, but they are quite huge and should be rewritten in this case, so I propose to keep both for now.

Reply:
The new name of this file is member_swap_optional.pass.cpp

Reply:
I have checked your proposal: 
```C++
X(X&& x) = delete
X& operator=(X&&) = delete;
```
In this case we have a lot of compile errors:
```C++
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(130,35): error : static assertion failed due to requirement 'noexcept(opt1.swap(opt2)) == true'
    130 |                     static_assert(noexcept(opt1.swap(opt2)) == true);
        |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(140,35): error : static assertion failed due to requirement 'noexcept(opt1.swap(opt2)) == true'
    140 |                     static_assert(noexcept(opt1.swap(opt2)) == true);
        |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(152,35): error : static assertion failed due to requirement 'noexcept(opt1.swap(opt2)) == true'
    152 |                     static_assert(noexcept(opt1.swap(opt2)) == true);
        |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(164,35): error : static assertion failed due to requirement 'noexcept(opt1.swap(opt2)) == true'
    164 |                     static_assert(noexcept(opt1.swap(opt2)) == true);
        |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp:18:
  In file included from C:\Work\github-oneapi\include\oneapi\dpl\optional:20:
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(350,23): error : static assertion failed due to requirement 'is_move_constructible_v<X>': optional<T>::swap requires T to be move constructible (N4950 [optional.swap]/1).
    350 |         static_assert(is_move_constructible_v<_Ty>,
        |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(133,26): note: in instantiation of member function 'std::optional<X>::swap' requested here
    133 |                     opt1.swap(opt2);
        |                          ^
  In file included from C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp:18:
  In file included from C:\Work\github-oneapi\include\oneapi\dpl\optional:20:
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(362,34): error : too many arguments to function call, expected single argument '_Right', have 2 arguments
    362 |                     swap(**this, *_Right); // intentional ADL
        |                     ~~~~         ^~~~~~~
  C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(348,23): note: 'swap' declared here
    348 |     _CONSTEXPR20 void swap(optional& _Right) noexcept(
        |                       ^    ~~~~~~~~~~~~~~~~
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(133,26): error : SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    133 |                     opt1.swap(opt2);
        |                          ^
  C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(348,23): note: 'swap' declared here
    348 |     _CONSTEXPR20 void swap(optional& _Right) noexcept(
        |                       ^
  C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(78,47): note: called by 'operator()'
     78 |             cgh.single_task<class KernelTest>([=]() {
        |                                               ^
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(144,26): error : SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    144 |                     opt1.swap(opt2);
        |                          ^
  C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(348,23): note: 'swap' declared here
    348 |     _CONSTEXPR20 void swap(optional& _Right) noexcept(
        |                       ^
  C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(78,47): note: called by 'operator()'
     78 |             cgh.single_task<class KernelTest>([=]() {
        |                                               ^
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(156,26): error : SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    156 |                     opt1.swap(opt2);
        |                          ^
  C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(348,23): note: 'swap' declared here
    348 |     _CONSTEXPR20 void swap(optional& _Right) noexcept(
        |                       ^
  C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(78,47): note: called by 'operator()'
     78 |             cgh.single_task<class KernelTest>([=]() {
        |                                               ^
C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(169,26): error : SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    169 |                     opt1.swap(opt2);
        |                          ^
  C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\optional(348,23): note: 'swap' declared here
    348 |     _CONSTEXPR20 void swap(optional& _Right) noexcept(
        |                       ^
  C:\Work\github-oneapi\test\xpu_api\utilities\optional\optional.object\optional.object.swap\member_swap_optional.pass.cpp(78,47): note: called by 'operator()'
  
     78 |             cgh.single_task<class KernelTest>([=]() {
  
        |                                               ^
  
  10 errors generated.
```

Reply:
Also I should note that [std::swap](https://en.cppreference.com/w/cpp/algorithm/swap) is applicable for both cases: 
```
Type requirements
-T must meet the requirements of CopyConstructible and CopyAssignable (until C++11)MoveConstructible and MoveAssignable (since C++11)
-T2 must meet the requirements of Swappable.
```
From my point of view this mean this test make sense in current state too.

Reply:
Fixed